web/src/main/java/org/cbioportal/web/StudyViewController.java

```diff
@@ -571,6 +574,271 @@ private static double parseValueLinear(ClinicalData c) {
+            studyIdsOfPatients = patients.stream().map(Patient::getCancerStudyIdentifier).collect(Collectors.toList());
+            patientToSamples = samples.stream().collect(
+                Collectors.groupingBy(Sample::getPatientStableId, Collectors.groupingBy(Sample::getCancerStudyIdentifier))
+            );
+        }
+
+        List<ClinicalData> clinicalDataList = clinicalDataFetcher.fetchClinicalData(
+            studyIds, sampleIds, patientIds, studyIdsOfPatients, sampleAttributeIds, patientAttributeIds, null
+        );
+
+        List<ClinicalData> sampleClinicalDataList;
+        // put all clinical data into sample form
+        if (CollectionUtils.isNotEmpty(patientAttributeIds)) {
+            sampleClinicalDataList = new ArrayList<>();
+            for (ClinicalData d: clinicalDataList) {
+                if (d.getSampleId() == null) {
+                    // null sample id means its a patient data, 
+                    //  we need to distribute the value to samples
+                    List<Sample> samplesForPatient = patientToSamples.get(
+                        d.getPatientId()
+                    ).get(d.getStudyId());
+                    if (samplesForPatient != null) {
+                        for (Sample s: samplesForPatient) {
+                            ClinicalData newData = new ClinicalData();
+                            newData.setAttrId(d.getAttrId());
+                            newData.setPatientId(d.getPatientId());
+                            newData.setStudyId(d.getStudyId());
+                            newData.setAttrValue(d.getAttrValue());
+                            newData.setSampleId(s.getStableId());
+                            sampleClinicalDataList.add(newData);
+                        }
+                    }
+                } else {
+                    // if its a sample data, just add it to the list
+                    sampleClinicalDataList.add(d);
+                }
+            }
+        } else {
+            sampleClinicalDataList = clinicalDataList;
+        }
+
+        // clinicalDataMap is a map sampleId->studyId->data
+        Map<String, Map<String, List<ClinicalData>>> clinicalDataMap = sampleClinicalDataList.stream().collect(Collectors.groupingBy(ClinicalData::getSampleId,
+            Collectors.groupingBy(ClinicalData::getStudyId)));
+        
+        // Group data by category
+        Map<String, List<ClinicalData>> groupedDetailedData = new HashMap<>();
+        boolean useLogScale = logScale && StudyViewController.isLogScalePossibleForAttribute(numericalAttributeId);
+        clinicalDataMap.forEach((k, v) -> v.forEach((m, n) -> {
+            if (
+                // n.size() == 2 means we have clinical data for the sample for both of the queried attributes
+                n.size() == 2 &&
+                    // check if the second data is numerical
+                    NumberUtils.isCreatable(n.get(1).getAttrValue())) {
```

Comment:
I don't think we really need the second comment, which lets us put that first comment above the loop, and clean up the indentation.
```suggestion
            // n.size() == 2 means we have clinical data for the sample for both of the queried attributes
            if (n.size() == 2 && NumberUtils.isCreatable(n.get(1).getAttrValue())) {                    
```